File: python/bob/bob.py
import string


def response(hey_bob: str) -> str:
    hey_bob = hey_bob.strip()
    is_shout = hey_bob.isupper()
    is_question = hey_bob.endswith('?')
    if not hey_bob:
        return 'Fine. Be that way!'

    if is_question and is_shout:
        return 'Calm down, I know what I\'m doing!'

    if is_question:
        return 'Sure.'

    if is_shout:
        return 'Whoa, chill out!'

    return 'Whatever.'

File: cpp/bob/bob.cpp
#include "bob.h"
#include <algorithm>
#include <cctype>

namespace bob {

std::string_view hey(std::string_view text) {
    bool isEmpty = true;
    bool hasAlpha = false;
    bool hasLower = false;
    char lastNonSpace = '\0';

    for (char c : text) {
        if (std::isspace(static_cast<unsigned char>(c))) {
            continue;
        }

        isEmpty = false;
        lastNonSpace = c;

        if (!std::isalpha(static_cast<unsigned char>(c))) {
            continue;
        }

        hasAlpha = true;
        if (std::islower(static_cast<unsigned char>(c))) {
            hasLower = true;
        }
    }

    if (isEmpty) {
        return "Fine. Be that way!";
    }

    const bool isQuestion = (lastNonSpace == '?');
    const bool isYell = hasAlpha && !hasLower;

    if (isQuestion && isYell) {
        return "Calm down, I know what I'm doing!";
    }

    if (isQuestion) {
        return "Sure.";
    }

    if (isYell) {
        return "Whoa, chill out!";
    }

    return "Whatever.";
}

}  // namespace bob
